File: HealthCareAPI/Services/QuestionService.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using HealthCareAPI.DTOs;
using HealthCareAPI.Entities;
using HealthCareAPI.Repositories;
using HealthCareAPI.Enum;

namespace HealthCareAPI.Services
{
    public class QuestionService
    {
        private readonly IUnitOfWork _unitOfWork;

        public QuestionService(IUnitOfWork unitOfWork)
        {
            _unitOfWork = unitOfWork;
        }

        public async Task<IEnumerable<QuestionDTO>> GetAllQuestionsAsync()
        {
            var questions = await _unitOfWork.QuestionRepository.GetQuestionsWithAccountAsync();
            return questions.Select(q => new QuestionDTO
            {
                QuestionID = q.QuestionID,
                AccountID = q.AccountID,
                Title = q.Title,
                Description = q.Description,
                Status = q.Status,
                CreatedAt = q.CreatedAt,
                UpdatedAt = q.UpdatedAt,
                Answer = q.Answer,
                Account = new AccountViewDTO
                {
                    FirstName = q.Account.FirstName,
                    LastName = q.Account.LastName,
                    DateOfBirth = q.Account.DateOfBirth,
                    Email = q.Account.Email
                }
            });
        }

        public async Task<bool> UpdateQuestionAsync(int id, QuestionDTO dto)
        {
            var question = await _unitOfWork.QuestionRepository.GetByIdAsync(id);
            if (question == null) return false;

            // Cập nhật các trường (tùy ý, có thể kiểm tra null hoặc cập nhật tất cả)
            question.Title = dto.Title;
            question.Description = dto.Description;
            question.Status = dto.Status;
            question.Answer = dto.Answer;
            question.UpdatedAt = DateTime.UtcNow;

            // Nếu muốn update các trường khác, bổ sung tại đây

            _unitOfWork.QuestionRepository.Update(question);
            await _unitOfWork.CompleteAsync();
            return true;
        }

        public async Task<NewQuestionDTO> CreateQuestionAsync(NewQuestionDTO dto)
        {
            var entity = new Question
            {
                AccountID = dto.AccountID,
                Title = dto.Title,
                Description = dto.Description,
                Status = StatusType.Submitted,
                CreatedAt = DateTime.UtcNow
            };

            await _unitOfWork.QuestionRepository.AddAsync(entity);
            await _unitOfWork.CompleteAsync();

            dto.QuestionID = entity.QuestionID;
            dto.Status = entity.Status;
            dto.CreatedAt = entity.CreatedAt;
            return dto;
        }
    }
}

File: HealthCareUI/src/services/question.service.ts
// services/question.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../app/app.config';

export interface QuestionTableDTO {
  questionID: number;
  accountID: string;
  title: string;
  description: string;
  status: number;
  createdAt: string;
  updatedAt?: string;
  answer?: string;
  account: {
    firstName: string;
    lastName: string;
    dateOfBirth: string;
    email: string;
  };
}
export interface QuestionDTO {
  questionID: number;
  title: string;
  description: string;
  status: number; // 0 - 4
  createdAt: string;
  updatedAt?: string;
  answer?: string;
}
export interface NewQuestionDTO {
  questionID: number;
  accountID: string;
  title: string;
  description: string;
  status: number;
  createdAt: string;
}

@Injectable({
  providedIn: 'root',
})
export class QuestionService {
  private apiUrl = environment.apiUrl + '/api/question'; // Đổi thành API thực tế của cậu

  constructor(private http: HttpClient) {}

  getAllQuestions(): Observable<QuestionTableDTO[]> {
    return this.http.get<QuestionTableDTO[]>(this.apiUrl);
  }

  // getQuestionById(id: number): Observable<QuestionTableDTO> {
  //   return this.http.get<QuestionTableDTO>(`${this.apiUrl}/${id}`);
  // }

  updateQuestion(id: number, question: QuestionTableDTO): Observable<any> {
    return this.http.put(`${this.apiUrl}/${id}`, question);
  }

  createQuestion(question: {
    accountID: string;
    title: string;
    description: string;
  }): Observable<NewQuestionDTO> {
    return this.http.post<NewQuestionDTO>(this.apiUrl, question);
  }

  getQuestionsByAccountId(accountId: string): Observable<QuestionTableDTO[]> {
    return this.http.get<QuestionTableDTO[]>(
      `${this.apiUrl}/account/${accountId}`
    );
  }

  // Có thể bổ sung các hàm khác như getById, update, delete nếu cần
}
